Google.Api.Gax.Grpc.Tests/GcpCallInvokerTests.cs

```diff
@@ -0,0 +1,246 @@
+            var bindChannel = fakeAdapter.CreatedChannels.Last();
+
+            // Act 2: Unbind
+            await invoker.AsyncUnaryCall(s_unbindMethod, null, default, unbindRequest).ResponseAsync;
+
+            // Assert
+            Assert.Single(fakeAdapter.CreatedChannels);
+            Assert.Equal(s_bindMethod, bindChannel.CallInvoker.Calls[0].Method);
+            Assert.Equal(s_unbindMethod, bindChannel.CallInvoker.Calls[1].Method);
+        }
+    }
+
+    public class FakeGrpcAdapter : GrpcAdapter
+    {
+        public List<FakeChannel> CreatedChannels { get; } = new List<FakeChannel>();
+
+        public FakeGrpcAdapter() : base(ApiTransports.Grpc) { }
+
+        private protected override ChannelBase CreateChannelImpl(ServiceMetadata serviceMetadata, string endpoint, ChannelCredentials credentials, GrpcChannelOptions options)
+        {
+            var channel = new FakeChannel(CreatedChannels.Count.ToString());
+            CreatedChannels.Add(channel);
+            return channel;
+        }
+    }
+
+    public class FakeChannel : ChannelBase
+    {
+        public string Id { get; }
+        public FakeCallInvoker CallInvoker { get; }
+
+        public FakeChannel(string id) : base("test-" + id)
+        {
+            Id = id;
+            CallInvoker = new FakeCallInvoker(this);
+        }
+
+        public override CallInvoker CreateCallInvoker() => CallInvoker;
+    }
+
+    public class FakeCallInvoker : CallInvoker
+    {
+        public FakeChannel Channel { get; }
+        public List<(Method<SimpleRequest, SimpleResponse> Method, SimpleRequest Request)> Calls { get; } = new List<(Method<SimpleRequest, SimpleResponse>, SimpleRequest)>();
+
+        public FakeCallInvoker(FakeChannel channel)
+        {
+            Channel = channel;
+        }
+
+        public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
+        {
+            if (method is Method<SimpleRequest, SimpleResponse> srMethod && request is SimpleRequest srRequest)
+            {
+                Calls.Add((srMethod, srRequest));
+
+                SimpleResponse response = new SimpleResponse();
+                // For Bind methods, the key is extracted from the response's "name" field.
+                // We use the request's "name" to populate this for the test.
+                response.Name = srRequest.Name;
+
+                if (srRequest.Name.Equals(GcpCallInvokerTest.IntroduceDelayKey))
+                {
+                    // This is to aid new channel creations for different RPCs having different affinity key values
+                    // If we do not do this, the channel gets freed up and reused in the test (which is what we want for most of the tests)
+
+                    // TODO: Substitute Thread.Sleep with FakeScheduler.Delay
+                    Thread.Sleep(5000);
```

Comment:
OK, let's change this from `Thread.Sleep` (I'm not comfortable with having that in a test to be honest)  for `Task.Delay`. We can't do that right here, we need to pass the delay to `CreateAsyncUnaryCall` and there see my comment.